compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2024 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.compose.ui.window
+
+import androidx.compose.runtime.Immutable
+import androidx.compose.ui.unit.Dp
+import androidx.compose.ui.unit.dp
+import androidx.compose.ui.window.WindowDecoration.SystemDefault
+import androidx.compose.ui.window.WindowDecoration.Undecorated
+
+/**
+ * Defines the options for window decoration.
+ */
+interface WindowDecoration {
+
+    /**
+     * Specifies that the default system decoration should be used.
+     */
+    object SystemDefault : WindowDecoration
+
+    /**
+     * Specifies that the window should be undecorated.
+     *
+     * If it is resizable, the given thickness will be used for the edge resizers.
+     */
+    @Immutable
+    class Undecorated(val resizerThickness: Dp = DefaultWindowResizerThickness) : WindowDecoration {
+        override fun equals(other: Any?): Boolean {
+            if (other !is Undecorated) return false
+            return other.resizerThickness == resizerThickness
+        }
+
+        override fun hashCode(): Int {
+            return resizerThickness.hashCode()
+        }
+    }
+}
+
+/**
+ * The default thickness of the resizers in an undecorated window.
+ */
+val DefaultWindowResizerThickness: Dp = 8.dp
```

Comment:
I'd move it from global scope to companion object

Reply:
> Agree with @igordmn - it should be sealed, but not class, sealed interface should work here. I guess it's covered by some policy - need to look for formal rule there, but the idea is to avoid implementation outside of our code. Explicit requirement in case of adding is pro, not cons

If we want a `sealed` interface here, we need to hide the actual subtypes from the user, to make the `when` scenario I mentioned impossible.

Reply:
> I'd move it from global scope to companion object

We can reuse AOSP pattern:

```
object WindowDecorationDefaults {
    val ResizerThickness: Dp = 8.dp
}
```

Reply:
On the other hand, it is only for `Undecorated`.

I delegate the decision to @m-sasha

Reply:
> we need to hide the actual subtypes from the user, to make the when scenario I mentioned impossible.

Why do we need to hide it? Why do we want to make such switch/when impossible?

Reply:
Because adding a new subtype will break it.

Reply:
> Explicit requirement in case of adding is pro, not cons

It's OK to require extra case handling in this case. It won't break _binary_ compatibility that we **have to** maintain

Reply:
I'd rather not dive into the exact semantics of what **binary** compatibility means.

De-facto, valid code that was written and compiled before the addition of another subtype will crash at runtime afterwards. That's breaking existing code.

Reply:
It's not source compatibility that will be broken. Existing compiled code will throw an exception at runtime.

Reply:
All of it is about possibilities and potential changes. Not a blocker at the moment I guess, we can discuss it when we'll really need this